File: src/sprite/Sprite.ts
import {Animator, Atlas, Integer, Milliseconds} from 'aseprite-atlas'
import {AtlasID} from '../atlas/AtlasID'
import {DecamillipixelXY, XY} from '../math/XY'
import {Layer} from './Layer'
import {ReadonlyRect, Rect} from '../math/Rect'
import {SpriteComposition} from './SpriteComposition'
import {WH} from '../math/WH'

/** A mapping from an immutable source atlas subtexture to a mutable rendered
    destination region. The sprite may be animated. */
export class Sprite {
  /** The atlas subtexture identifier. The identifier should be unique within
      the atlas but may be referenced by multiple sprites. */
  private readonly _id: AtlasID

  /** See SpriteComposition. */
  private _constituentID: AtlasID
  private _composition: SpriteComposition

  /** The position and size of the destination, includes scaling. The width and
      height are unsigned and never negative even when scaling indicates the
      sprite will be flipped when rendered.

      Each animation cel has the same size so an sprite's dimensions only change
      when instructed by the caller. Specifying a different destination width or
      height than the source truncates or repeats the scaled rendered source.

      Sprite.bounds are used to determine when the sprite is on screen and
      therefor should be drawn, as well as for rendering itself.

      Sprite.bounds are used for some collision tests (see CollisionPredicate)
      as entity bounds are calculated in part by Sprite.bounds,
      SpriteRect.bounds, and Entity.bounds. */
  private readonly _bounds: Rect

  /** The drawing layer. A higher number indicates a later draw relative other
      sprites. The effect is that an sprite may be drawn above or below other
      sprites in a controlled manner allowing for effects like shadows to
      consistently be drawn below the object meant to cast it. For objects on
      the same layer, the greater destination position and height is drawn
      last. */
  private _layer: Layer

  private readonly _scale: XY

  private readonly _animator: Animator

  /** Specifies the initial marquee offset. Marquee offsets are all relative
      relative the absolute game time and therefor synchronized by default. */
  private readonly _wrap: DecamillipixelXY

  /** Specifies the additional marquee offset added by the shader according to
      the game clock. */
  private readonly _wrapVelocity: DecamillipixelXY

  /** Optionally, compute the sprite size from the animation source's size and
      the specified scale. */
  static withAtlasSize(atlas: Atlas, props: Sprite.Props): Sprite {
    const animation = atlas.animations[props.id]
    const size =
      props.size ??
      new WH(
        props.w ?? animation.size.w * Math.abs(props.scale?.x ?? props.sx ?? 1),
        props.h ?? animation.size.h * Math.abs(props.scale?.y ?? props.sy ?? 1)
      )
    return new Sprite({...props, size})
  }

  constructor(props: Sprite.Props) {
    this._id = props.id
    this._constituentID = props.constituentID ?? props.id
    this._composition = props.composition ?? SpriteComposition.SOURCE
    this._scale = props.scale ?? new XY(props.sx ?? 1, props.sy ?? 1)
    Sprite.validateScale(this._scale)
    this._bounds = props.bounds ?? {
      position: props.position ?? new XY(props.x ?? 0, props.y ?? 0),
      size: props.size ?? new WH(props.w ?? 0, props.h ?? 0)
    }
    this._layer = props.layer ?? Layer.DEFAULT
    this._animator = props.animator ?? {
      period: props.period ?? 0,
      exposure: props.exposure ?? 0
    }
    this._wrap = props.wrap ?? new XY(props.wx ?? 0, props.wy ?? 0)
    this._wrapVelocity =
      props.wrapVelocity ?? new XY(props.wvx ?? 0, props.wvy ?? 0)
  }

  get id(): AtlasID {
    return this._id
  }

  get constituentID(): AtlasID {
    return this._constituentID
  }

  set constituentID(id: AtlasID) {
    this._constituentID = id
  }

  get composition(): SpriteComposition {
    return this._composition
  }

  set composition(composition: SpriteComposition) {
    this._composition = composition
  }

  get bounds(): ReadonlyRect {
    return this._bounds
  }

  moveBy(by: Readonly<XY>): void {
    this._bounds.position.x += by.x
    this._bounds.position.y += by.y
  }

  moveTo(to: Readonly<XY>): void {
    this._bounds.position.x = to.x
    this._bounds.position.y = to.y
  }

  sizeTo(unscaledTo: Readonly<WH>): void {
    this._bounds.size.w = unscaledTo.w * Math.abs(this.scale.x)
    this._bounds.size.h = unscaledTo.h * Math.abs(this.scale.y)
  }

  get layer(): Layer {
    return this._layer
  }

  set layer(layer: Layer) {
    this._layer = layer
  }

  /** See Sprite.compareElevation(). */
  compareElevation(sprite: Readonly<Sprite>): Integer {
    return Sprite.compareElevation(this, sprite)
  }

  get scale(): Readonly<XY> {
    return this._scale
  }

  /** Set absolute scale. */
  scaleTo(to: Readonly<XY>): void {
    Sprite.validateScale(to)
    this._bounds.size.w *= Math.abs(to.x / this.scale.x)
    this._bounds.size.h *= Math.abs(to.y / this.scale.y)
    this._scale.x = to.x
    this._scale.y = to.y
  }

  /** Multiply by scale. */
  scaleBy(by: Readonly<XY>): void {
    Sprite.validateScale(by)
    this._bounds.size.w *= Math.abs(by.x)
    this._bounds.size.h *= Math.abs(by.y)
    this._scale.x *= by.x
    this._scale.y *= by.y
  }

  get animator(): Readonly<Animator> {
    return this._animator
  }

  animate(atlas: Atlas, time: Milliseconds): void {
    const {period, exposure} = Animator.animate(
      this.animator.period,
      this.animator.exposure + time,
      atlas.animations[this.id]
    )
    this._animator.period = period
    this._animator.exposure = exposure
  }

  resetAnimation(): void {
    this._animator.period = 0
    this._animator.exposure = 0
  }

  animation(atlas: Atlas): Atlas.Animation {
    return atlas.animations[this.id]
  }

  cel(atlas: Atlas): Atlas.Cel {
    return this.animation(atlas).cels[this.celIndex(atlas)]
  }

  bodies(atlas: Atlas): Rect[] {
    return this.cel(atlas).slices.map(({x, y, w, h}) => Rect.make(x, y, w, h))
  }

  celIndex(atlas: Atlas): Integer {
    return Animator.index(this.animator.period, this.animation(atlas).cels)
  }

  get wrap(): Readonly<XY> {
    return this._wrap
  }

  wrapTo(to: Readonly<XY>): void {
    this._wrap.x = to.x
    this._wrap.y = to.y
  }

  get wrapVelocity(): Readonly<XY> {
    return this._wrapVelocity
  }

  wrapVelocityTo(to: Readonly<XY>): void {
    this._wrapVelocity.x = to.x
    this._wrapVelocity.y = to.y
  }
}

export namespace Sprite {
  /** Outermost references have precedence over destructured properties when
      specified. E.g., bounds > position > x. Callers can mutate the referenced
      objects at their own peril. */
  export interface Props {
    readonly id: AtlasID

    /** Defaults to Props.id. */
    readonly constituentID?: AtlasID
    /** Defaults to SpriteComposition.SOURCE. */
    readonly composition?: SpriteComposition

    /** Defaults to a rectangle with atlas source width and height at 0, 0.
        If bounds or size are specified, the source size is not considered. */
    readonly bounds?: Rect
    readonly position?: XY
    readonly x?: Integer
    readonly y?: Integer
    readonly size?: WH
    readonly w?: Integer
    readonly h?: Integer

    /** Defaults to zero. */
    readonly layer?: Integer

    /** Defaults to 1, 1. */
    readonly scale?: XY
    readonly sx?: Integer
    readonly sy?: Integer

    /** Defaults to {period: 0, exposure: 0}. */
    readonly animator?: Animator
    readonly period?: Integer
    readonly exposure?: Milliseconds

    /** Defaults to 0, 0. */
    readonly wrap?: DecamillipixelXY
    readonly wx?: Integer
    readonly wy?: Integer

    /** Defaults to 0, 0. */
    readonly wrapVelocity?: DecamillipixelXY
    readonly wvx?: Integer
    readonly wvy?: Integer
  }

  /** Return the draw order. */
  export function compareElevation(
    lhs: Readonly<Sprite>,
    rhs: Readonly<Sprite>
  ): Integer {
    return (
      lhs.layer - rhs.layer ||
      lhs.bounds.position.y +
        lhs.bounds.size.h -
        (rhs.bounds.position.y + rhs.bounds.size.h)
    )
  }

  export function validateScale({x, y}: Readonly<XY>): void {
    if (!x || !y) throw new Error(`Scale must be nonzero (${x}, ${y}).`)
  }
}

File: src/sprites/sprite.rs
use super::{SpriteComposition, SpriteLayer};
use crate::atlas::{AnimationID, Animator, Atlas};
use crate::math::{Millis, R16, WH16, XY, XY16};
use std::convert::TryFrom;
use std::{convert::TryInto, num::NonZeroI16};

/// A mapping from an Atlas Animation to a level region. This includes all
/// distinct state needed to represent an instance to the shader.
pub struct Sprite {
  /// Source Animation state.
  animator: Animator,
  /// The source Animation subtextures identifer to draw from.
  id: AnimationID,
  /// The Animation subtextures identifer to compose with source. The
  /// constituent uses the source Animator's period and as such cannot be
  /// animated distinctly. Wrapping is not applied to the constituent.
  constituent: AnimationID,
  /// The composition operator for source and constituent.
  composition: SpriteComposition,
  /// The level region to draw to including scale. The width and height are
  /// unsigned and never negative even when scaling indicates the sprite will be
  /// flipped when rendered.
  ///
  /// Each animation cel has the same size so an sprite's dimensions only change
  /// when explicitly instructed. Specifying a different destination width or
  /// height than the source truncates or repeats the scaled rendered source.
  ///
  /// This region is used to determine when the sprite is on screen and therefor
  /// should be drawn, as well as for rendering itself. The region is also used
  /// for certain collision tests.
  bounds: R16,
  /// The destination multiple.
  scale: XY<NonZeroI16>,
  /// The initial marquee translation offset in .1 pixels. The total offset is
  /// calculated using the global game clock and therefor synchronized by
  /// default except where this property deviates.
  wrap: XY16,
  /// The marquee translation speed in .1 pixels per second (or 1 px / 10 000 ms
  /// or 1 px / 10 s). The offset for each render is calculated by the shader as
  /// `wrap + wrap_velocity * game_clock`.
  wrap_velocity: XY16,
  /// The painting draw order.
  layer: u8,
}

impl Sprite {
  pub fn new(
    animator: Animator,
    id: AnimationID,
    constituent: AnimationID,
    composition: SpriteComposition,
    bounds: R16,
    scale: XY<NonZeroI16>,
    wrap: XY16,
    wrap_velocity: XY16,
    layer: SpriteLayer,
  ) -> Self {
    Self {
      animator,
      id,
      constituent,
      composition,
      bounds,
      scale,
      wrap,
      wrap_velocity,
      layer: layer as u8,
    }
  }

  pub fn animate(&mut self, atlas: &Atlas, exposure: Millis) {
    self.animator.animate(&atlas.animations[&self.id], exposure);
  }

  pub fn reset_animation(&mut self) {
    self.animator.reset();
  }

  pub fn get_id(&self) -> AnimationID {
    self.id
  }

  pub fn get_constituent(&self) -> AnimationID {
    self.constituent
  }

  pub fn get_composition(&self) -> SpriteComposition {
    self.composition
  }

  pub fn get_bounds(&self) -> &R16 {
    &self.bounds
  }

  pub fn move_by(&mut self, by: &XY16) {
    self.bounds = self.bounds.clone() + by.clone();
  }

  pub fn move_to(&mut self, to: &XY16) {
    self.bounds = self.bounds.move_to(to);
  }

  pub fn size_to(&mut self, unscaled_to: &WH16) {
    let scale: XY16 = self.scale.clone().into();
    self.bounds.to =
      self.bounds.from.clone() + (unscaled_to.clone() * scale.abs()).into();
  }

  pub fn get_scale(&self) -> &XY<NonZeroI16> {
    &self.scale
  }

  pub fn scale_by(&mut self, by: &XY<NonZeroI16>) {
    let by: XY16 = by.clone().into();
    self.bounds.to = self.bounds.from.clone()
      + (self.bounds.to.clone() - self.bounds.from.clone()) * by.abs();
    let scale: XY16 = self.scale.clone().into();
    self.scale = (scale * by).try_into().expect("Scalar is zero.");
  }

  pub fn scale_to(&mut self, to: &XY<NonZeroI16>) {
    let scale: XY<f32> = self.scale.clone().into();
    self.scale = to.clone();
    let wh: XY<f32> =
      (self.bounds.to.clone() - self.bounds.from.clone()).into();
    let scaled_wh = wh * (XY::<f32>::from(to.clone()) / scale).abs();
    self.bounds.to = self.bounds.from.clone()
      + scaled_wh.try_into().expect("WH f32 to i16 conversion failed.");
  }

  pub fn get_wrap(&self) -> &XY16 {
    &self.wrap
  }

  pub fn wrap_to(&mut self, to: &XY16) {
    self.wrap = to.clone();
  }

  pub fn get_wrap_velocity(&self) -> &XY16 {
    &self.wrap_velocity
  }

  pub fn wrap_velocity_to(&mut self, to: &XY16) {
    self.wrap_velocity = to.clone();
  }

  pub fn get_layer(&self) -> u8 {
    self.layer
  }

  pub fn elevate_layer_to(&mut self, layer: SpriteLayer) {
    self.layer = layer as u8;
  }

  pub fn elevate_layer_by(&mut self, layer: SpriteLayer) {
    self.layer += layer as u8;
  }

  pub fn serialize(
    &self,
    config: &bincode::Config,
    atlas: &Atlas,
  ) -> bincode::Result<Vec<u8>> {
    let mut bytes = Vec::new();

    bytes.append(
      &mut config.serialize(
        &self
          .animator
          .cel(&atlas.animations[&self.id])
          .ok_or(bincode::ErrorKind::Custom("No source Cel.".to_string()))?
          .bounds,
      )?,
    );
    bytes.append(
      &mut config.serialize(
        &self
          .animator
          .cel(&atlas.animations[&self.constituent])
          .ok_or(bincode::ErrorKind::Custom("No constituent Cel.".to_string()))?
          .bounds,
      )?,
    );
    bytes.append(&mut config.serialize(&self.composition)?);
    bytes.append(&mut config.serialize(&self.bounds)?);
    bytes.append(&mut config.serialize(&self.scale)?);
    bytes.append(&mut config.serialize(&self.wrap)?);
    bytes.append(&mut config.serialize(&self.wrap_velocity)?);

    Ok(bytes)
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::atlas::{Animation, Cel, Playback};
  use crate::math::WH;
  use std::collections::hash_map::HashMap;

  #[test]
  fn scale() {
    let mut sprite = Sprite::new(
      Animator::new(0, 0.),
      AnimationID::Bee,
      AnimationID::Bee,
      SpriteComposition::Source,
      R16::new_wh(1, 2, 3, 4),
      XY::cast_into_non_zero(1, 1).unwrap(),
      XY::new(0, 0),
      XY::new(0, 0),
      SpriteLayer::Default,
    );

    assert_eq!(sprite.get_scale(), &XY::cast_into_non_zero(1, 1).unwrap());

    sprite.scale_by(&XY::cast_into_non_zero(2, 2).unwrap());
    assert_eq!(sprite.get_scale(), &XY::cast_into_non_zero(2, 2).unwrap());
    assert_eq!(sprite.get_bounds(), &R16::new_wh(1, 2, 6, 8));

    sprite.scale_to(&XY::cast_into_non_zero(3, 3).unwrap());
    assert_eq!(sprite.get_scale(), &XY::cast_into_non_zero(3, 3).unwrap());
    assert_eq!(sprite.get_bounds(), &R16::new_wh(1, 2, 9, 12));
  }

  #[test]
  fn wrap() {
    let mut sprite = Sprite::new(
      Animator::new(0, 0.),
      AnimationID::Bee,
      AnimationID::Bee,
      SpriteComposition::Source,
      R16::new(1, 2, 3, 4),
      XY::cast_into_non_zero(1, 1).unwrap(),
      XY::new(5, 6),
      XY::new(7, 8),
      SpriteLayer::Default,
    );

    assert_eq!(sprite.get_wrap(), &XY::new(5, 6));
    assert_eq!(sprite.get_wrap_velocity(), &XY::new(7, 8));

    sprite.wrap_to(&XY::new(9, 10));
    sprite.wrap_velocity_to(&XY::new(11, 12));
    assert_eq!(sprite.get_wrap(), &XY::new(9, 10));
    assert_eq!(sprite.get_wrap_velocity(), &XY::new(11, 12));
  }

  #[test]
  fn layer() {
    let mut sprite = Sprite::new(
      Animator::new(0, 0.),
      AnimationID::Bee,
      AnimationID::Bee,
      SpriteComposition::Source,
      R16::new(1, 2, 3, 4),
      XY::cast_into_non_zero(1, 1).unwrap(),
      XY::new(0, 0),
      XY::new(0, 0),
      SpriteLayer::Default,
    );

    assert_eq!(sprite.get_layer(), SpriteLayer::Default as u8);

    sprite.elevate_layer_by(SpriteLayer::UILo);
    assert_eq!(
      sprite.get_layer(),
      (SpriteLayer::Default as u8) + (SpriteLayer::UILo as u8)
    );

    sprite.elevate_layer_to(SpriteLayer::Default);
    assert_eq!(sprite.get_layer(), (SpriteLayer::Default as u8));
  }

  #[rustfmt::skip]
  #[test]
  fn serialize() {
    let mut animations = HashMap::new();
    let cels = vec![Cel {
      bounds: R16::new(7, 8, 12, 14),
      duration: 100.,
      slices: vec![],
    }];
    animations.insert(
      AnimationID::Bee,
      Animation {
        wh: WH::new(5, 6),
        cels,
        duration: 100.,
        direction: Playback::Forward,
      },
    );
    let atlas = Atlas {
      version: "1.2.3.4".to_string(),
      filename: "atlas.png".to_string(),
      format: "I8".to_string(),
      wh: WH::new(256, 256),
      animations,
    };
    let sprite = Sprite::new(
      Animator::new(0, 0.),
      AnimationID::Bee,
      AnimationID::Bee,
      SpriteComposition::SourceMask,
      R16::new(9, 10, 55, 72),
      XY::cast_into_non_zero(11, 13).unwrap(),
      XY::new(15, 16),
      XY::new(17, 18),
      SpriteLayer::UICursor,
    );
    let bytes =
      sprite.serialize(bincode::config().big_endian(), &atlas).unwrap();
    assert_eq!(
      bytes,
      vec![
        0, 7, 0, 8, 0, 12, 0, 14,
        0, 7, 0, 8, 0, 12, 0, 14,
        0, sprite.composition as u8,
        0, 9, 0, 10, 0, 55, 0, 72,
        0, 11, 0, 13,
        0, 15, 0, 16,
        0, 17, 0, 18
      ]
    );
  }
}
